File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/AuthenticateUserViewModel.java
package droiddevelopers254.droidconke.viewmodels;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseUser;

import droiddevelopers254.droidconke.datastates.AuthenticateUserState;
import droiddevelopers254.droidconke.repository.AuthenticateUserRepo;

public class AuthenticateUserViewModel extends ViewModel {
    private MediatorLiveData<AuthenticateUserState> userStateMediatorLiveData;
    private AuthenticateUserRepo authenticateUserRepo;

    public AuthenticateUserViewModel(){
        userStateMediatorLiveData= new MediatorLiveData<>();
        authenticateUserRepo= new AuthenticateUserRepo();
    }

    public LiveData<AuthenticateUserState> getAuthenticateResponse(){

        return userStateMediatorLiveData;
    }

    public void authenticateUser(FirebaseUser firebaseUser){
        final LiveData<AuthenticateUserState> stateLiveData=authenticateUserRepo.checkUserExistence(firebaseUser);
        userStateMediatorLiveData.addSource(stateLiveData,
                userStateMediatorLiveData->{
            if (this.userStateMediatorLiveData.hasActiveObservers()){
                this.userStateMediatorLiveData.removeSource(stateLiveData);
            }
            this.userStateMediatorLiveData.setValue(userStateMediatorLiveData);
                });
    }
}

File: app/src/main/java/droiddevelopers254/droidconke/viewmodels/AuthenticateUserViewModel.kt
package droiddevelopers254.droidconke.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseUser
import droiddevelopers254.droidconke.datastates.AuthenticateUserState
import droiddevelopers254.droidconke.repository.AuthenticateUserRepo

class AuthenticateUserViewModel : ViewModel() {
    private val userStateMediatorLiveData: MediatorLiveData<AuthenticateUserState> = MediatorLiveData()
    private val authenticateUserRepo: AuthenticateUserRepo = AuthenticateUserRepo()

    val authenticateResponse: LiveData<AuthenticateUserState>
        get() = userStateMediatorLiveData

    fun authenticateUser(firebaseUser: FirebaseUser) {
        val stateLiveData = authenticateUserRepo.checkUserExistence(firebaseUser)
        userStateMediatorLiveData.addSource(stateLiveData
        ) { userStateMediatorLiveData ->
            when {
                this.userStateMediatorLiveData.hasActiveObservers() -> this.userStateMediatorLiveData.removeSource(stateLiveData)
            }
            this.userStateMediatorLiveData.setValue(userStateMediatorLiveData)
        }
    }
}
